File: src/org/thoughtcrime/securesms/attachments/PointerAttachment.java
package org.thoughtcrime.securesms.attachments;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.thoughtcrime.securesms.database.AttachmentDatabase;
import org.thoughtcrime.securesms.stickers.StickerLocator;
import org.thoughtcrime.securesms.util.Base64;
import org.whispersystems.libsignal.util.guava.Optional;
import org.whispersystems.signalservice.api.messages.SignalServiceAttachment;
import org.whispersystems.signalservice.api.messages.SignalServiceDataMessage;

import java.util.LinkedList;
import java.util.List;

public class PointerAttachment extends Attachment {

  private PointerAttachment(@NonNull String contentType, int transferState, long size,
                            @Nullable String fileName,  @NonNull String location,
                            @Nullable String key, @Nullable String relay,
                            @Nullable byte[] digest, @Nullable String fastPreflightId, boolean voiceNote,
                            int width, int height, @Nullable String caption, @Nullable StickerLocator stickerLocator)
  {
    super(contentType, transferState, size, fileName, location, key, relay, digest, fastPreflightId, voiceNote, width, height, false, caption, stickerLocator);
  }

  @Nullable
  @Override
  public Uri getDataUri() {
    return null;
  }

  @Nullable
  @Override
  public Uri getThumbnailUri() {
    return null;
  }


  public static List<Attachment> forPointers(Optional<List<SignalServiceAttachment>> pointers) {
    List<Attachment> results = new LinkedList<>();

    if (pointers.isPresent()) {
      for (SignalServiceAttachment pointer : pointers.get()) {
        Optional<Attachment> result = forPointer(Optional.of(pointer));

        if (result.isPresent()) {
          results.add(result.get());
        }
      }
    }

    return results;
  }

  public static List<Attachment> forPointers(List<SignalServiceDataMessage.Quote.QuotedAttachment> pointers) {
    List<Attachment> results = new LinkedList<>();

    if (pointers != null) {
      for (SignalServiceDataMessage.Quote.QuotedAttachment pointer : pointers) {
        Optional<Attachment> result = forPointer(pointer);

        if (result.isPresent()) {
          results.add(result.get());
        }
      }
    }

    return results;
  }

  public static Optional<Attachment> forPointer(Optional<SignalServiceAttachment> pointer) {
    return forPointer(pointer, null, null);
  }

  public static Optional<Attachment> forPointer(Optional<SignalServiceAttachment> pointer, @Nullable StickerLocator stickerLocator) {
    return forPointer(pointer, stickerLocator, null);
  }

  public static Optional<Attachment> forPointer(Optional<SignalServiceAttachment> pointer, @Nullable StickerLocator stickerLocator, @Nullable String fastPreflightId) {
    if (!pointer.isPresent() || !pointer.get().isPointer()) return Optional.absent();

    String encodedKey = null;

    if (pointer.get().asPointer().getKey() != null) {
      encodedKey = Base64.encodeBytes(pointer.get().asPointer().getKey());
    }

    return Optional.of(new PointerAttachment(pointer.get().getContentType(),
                                      AttachmentDatabase.TRANSFER_PROGRESS_PENDING,
                                      pointer.get().asPointer().getSize().or(0),
                                      pointer.get().asPointer().getFileName().orNull(),
                                      String.valueOf(pointer.get().asPointer().getId()),
                                      encodedKey, null,
                                      pointer.get().asPointer().getDigest().orNull(),
                                      fastPreflightId,
                                      pointer.get().asPointer().getVoiceNote(),
                                      pointer.get().asPointer().getWidth(),
                                      pointer.get().asPointer().getHeight(),
                                      pointer.get().asPointer().getCaption().orNull(),
        stickerLocator));

  }

  public static Optional<Attachment> forPointer(SignalServiceDataMessage.Quote.QuotedAttachment pointer) {
    SignalServiceAttachment thumbnail = pointer.getThumbnail();

    return Optional.of(new PointerAttachment(pointer.getContentType(),
                                             AttachmentDatabase.TRANSFER_PROGRESS_PENDING,
                                             thumbnail != null ? thumbnail.asPointer().getSize().or(0) : 0,
                                             pointer.getFileName(),
                                             String.valueOf(thumbnail != null ? thumbnail.asPointer().getId() : 0),
                                             thumbnail != null && thumbnail.asPointer().getKey() != null ? Base64.encodeBytes(thumbnail.asPointer().getKey()) : null,
                                             null,
                                             thumbnail != null ? thumbnail.asPointer().getDigest().orNull() : null,
                                             null,
                                             false,
                                             thumbnail != null ? thumbnail.asPointer().getWidth() : 0,
                                             thumbnail != null ? thumbnail.asPointer().getHeight() : 0,
                                             thumbnail != null ? thumbnail.asPointer().getCaption().orNull() : null,
                                             null));
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/attachments/PointerAttachment.kt
package org.thoughtcrime.securesms.attachments

import android.net.Uri
import android.os.Parcel
import androidx.annotation.VisibleForTesting
import org.signal.core.util.Base64
import org.thoughtcrime.securesms.blurhash.BlurHash
import org.thoughtcrime.securesms.database.AttachmentTable
import org.thoughtcrime.securesms.stickers.StickerLocator
import org.whispersystems.signalservice.api.InvalidMessageStructureException
import org.whispersystems.signalservice.api.messages.SignalServiceAttachment
import org.whispersystems.signalservice.api.util.AttachmentPointerUtil
import org.whispersystems.signalservice.internal.push.DataMessage
import java.util.Optional
import java.util.UUID

class PointerAttachment : Attachment {
  @VisibleForTesting
  constructor(
    contentType: String?,
    transferState: Int,
    size: Long,
    fileName: String?,
    cdn: Cdn,
    location: String,
    key: String?,
    iv: ByteArray?,
    digest: ByteArray?,
    incrementalDigest: ByteArray?,
    incrementalMacChunkSize: Int,
    fastPreflightId: String?,
    voiceNote: Boolean,
    borderless: Boolean,
    videoGif: Boolean,
    width: Int,
    height: Int,
    uploadTimestamp: Long,
    caption: String?,
    stickerLocator: StickerLocator?,
    blurHash: BlurHash?,
    uuid: UUID?
  ) : super(
    contentType = contentType,
    transferState = transferState,
    size = size,
    fileName = fileName,
    cdn = cdn,
    remoteLocation = location,
    remoteKey = key,
    remoteIv = iv,
    remoteDigest = digest,
    incrementalDigest = incrementalDigest,
    fastPreflightId = fastPreflightId,
    voiceNote = voiceNote,
    borderless = borderless,
    videoGif = videoGif,
    width = width,
    height = height,
    incrementalMacChunkSize = incrementalMacChunkSize,
    quote = false,
    uploadTimestamp = uploadTimestamp,
    caption = caption,
    stickerLocator = stickerLocator,
    blurHash = blurHash,
    audioHash = null,
    transformProperties = null,
    uuid = uuid
  )

  constructor(parcel: Parcel) : super(parcel)

  override val uri: Uri? = null
  override val publicUri: Uri? = null
  override val thumbnailUri: Uri? = null

  companion object {
    @JvmStatic
    fun forPointers(pointers: Optional<List<SignalServiceAttachment>>): List<Attachment> {
      if (!pointers.isPresent) {
        return emptyList()
      }

      return pointers.get()
        .map { forPointer(Optional.ofNullable(it)) }
        .filter { it.isPresent }
        .map { it.get() }
    }

    @JvmStatic
    @JvmOverloads
    fun forPointer(
      pointer: Optional<SignalServiceAttachment>,
      stickerLocator: StickerLocator? = null,
      fastPreflightId: String? = null,
      transferState: Int = AttachmentTable.TRANSFER_PROGRESS_PENDING
    ): Optional<Attachment> {
      if (!pointer.isPresent || !pointer.get().isPointer()) {
        return Optional.empty()
      }

      val encodedKey: String? = pointer.get().asPointer().key?.let { Base64.encodeWithPadding(it) }

      return Optional.of(
        PointerAttachment(
          contentType = pointer.get().contentType,
          transferState = transferState,
          size = pointer.get().asPointer().size.orElse(0).toLong(),
          fileName = pointer.get().asPointer().fileName.orElse(null),
          cdn = Cdn.fromCdnNumber(pointer.get().asPointer().cdnNumber),
          location = pointer.get().asPointer().remoteId.toString(),
          key = encodedKey,
          iv = null,
          digest = pointer.get().asPointer().digest.orElse(null),
          incrementalDigest = pointer.get().asPointer().incrementalDigest.orElse(null),
          incrementalMacChunkSize = pointer.get().asPointer().incrementalMacChunkSize,
          fastPreflightId = fastPreflightId,
          voiceNote = pointer.get().asPointer().voiceNote,
          borderless = pointer.get().asPointer().isBorderless,
          videoGif = pointer.get().asPointer().isGif,
          width = pointer.get().asPointer().width,
          height = pointer.get().asPointer().height,
          uploadTimestamp = pointer.get().asPointer().uploadTimestamp,
          caption = pointer.get().asPointer().caption.orElse(null),
          stickerLocator = stickerLocator,
          blurHash = BlurHash.parseOrNull(pointer.get().asPointer().blurHash.orElse(null)),
          uuid = pointer.get().asPointer().uuid
        )
      )
    }

    fun forPointer(quotedAttachment: DataMessage.Quote.QuotedAttachment): Optional<Attachment> {
      val thumbnail: SignalServiceAttachment? = try {
        if (quotedAttachment.thumbnail != null) {
          AttachmentPointerUtil.createSignalAttachmentPointer(quotedAttachment.thumbnail)
        } else {
          null
        }
      } catch (e: InvalidMessageStructureException) {
        return Optional.empty()
      }

      return Optional.of(
        PointerAttachment(
          contentType = quotedAttachment.contentType!!,
          transferState = AttachmentTable.TRANSFER_PROGRESS_PENDING,
          size = (if (thumbnail != null) thumbnail.asPointer().size.orElse(0) else 0).toLong(),
          fileName = quotedAttachment.fileName,
          cdn = Cdn.fromCdnNumber(thumbnail?.asPointer()?.cdnNumber ?: 0),
          location = thumbnail?.asPointer()?.remoteId?.toString() ?: "0",
          key = thumbnail?.asPointer()?.key?.let { Base64.encodeWithPadding(it) },
          iv = null,
          digest = thumbnail?.asPointer()?.digest?.orElse(null),
          incrementalDigest = thumbnail?.asPointer()?.incrementalDigest?.orElse(null),
          incrementalMacChunkSize = thumbnail?.asPointer()?.incrementalMacChunkSize ?: 0,
          fastPreflightId = null,
          voiceNote = false,
          borderless = false,
          videoGif = false,
          width = thumbnail?.asPointer()?.width ?: 0,
          height = thumbnail?.asPointer()?.height ?: 0,
          uploadTimestamp = thumbnail?.asPointer()?.uploadTimestamp ?: 0,
          caption = thumbnail?.asPointer()?.caption?.orElse(null),
          stickerLocator = null,
          blurHash = null,
          uuid = thumbnail?.asPointer()?.uuid
        )
      )
    }
  }
}
